src/nvim/ex_docmd.c

```diff
@@ -9624,21 +9624,41 @@ static void ex_folddo(exarg_T *eap)
 
 static void ex_terminal(exarg_T *eap)
 {
+  // We will call termopen() with shell as an arglist if not given a {cmd}.
   char *name = (char *)p_sh;  // Default to 'shell' if {cmd} is not given.
+  char ex_cmd[1024] = "";
   bool mustfree = false;
-  char *lquote = "['";
-  char *rquote = "']";
 
   if (*eap->arg != NUL) {
     name = (char *)vim_strsave_escaped(eap->arg, (char_u *)"\"\\");
     mustfree = true;
-    lquote = rquote = "\"";
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen(\"%s\") | startinsert",
+             eap->forceit ? "!" : "", name);
+  } else {
+    char **argv = build_argv((char *)p_sh);
+    char **p = argv;
+    char tempstring[512] = "";
+    char shell_argv[512] = "";
```

Comment:
this is an unusual way to initialize a char array, normally we do:

    char foo[512];

If you really want zero it out, do:

    char foo[512] = { 0 };